src/core/document.js

```diff
@@ -308,29 +307,112 @@ var Page = (function PageClosure() {
     },
 
     getAnnotationsData: function Page_getAnnotationsData(intent) {
-      var annotations = this.annotations;
-      var annotationsData = [];
-      for (var i = 0, n = annotations.length; i < n; ++i) {
-        if (!intent || isAnnotationRenderable(annotations[i], intent)) {
-          annotationsData.push(annotations[i].data);
+      var annotationsPromise = this.annotations;
+      return annotationsPromise.then(function (annotations) {
+        var annotationsData = [];
+        for (var i = 0, n = annotations.length; i < n; ++i) {
+          if (!intent || isAnnotationRenderable(annotations[i], intent)) {
+            annotationsData.push(annotations[i].data);
+          }
         }
-      }
-      return annotationsData;
+        return annotationsData;
+      });
     },
 
     get annotations() {
-      var annotations = [];
+      var AnnotationWorkerTask = (function AnnotationWorkerTaskClosure() {
```

Comment:
I think it would be better to import the existing `WorkerTask` class from `src/core/worker.js` for this instead since the code itself looks exactly the same.

Reply:
That still wouldn't address the underlying problem with `WorkerTask`s in this PR, since the purpose of `WorkerTask` is to allow waiting for the completion of various running tasks in `src/core/worker.js` to avoid e.g. intermittent issues in tests. Note how a worker won't be terminated until all running `WorkerTask`s have finished:
https://github.com/mozilla/pdf.js/blob/55e3f97aa97ad648ef955203524bbdb5593671d1/src/core/worker.js#L832-L846

As is, this PR (and the previous one) is completely breaking that assumption!

---

It seems to me that this PR still suffers from the same issue as the original one: It's simply *way* too large, with lots of different changes, making it very difficult to follow.
I'd thus suggest splitting this up a lot more, for example let's start with a PR that *only* makes the annotation parsing async, then one to parse the `AcroForm` dictionary[1], then one to use that dictionary when parsing annotations, and so on...

[1] Also, correct parsing of `AcroForm` dictionaries should also fix the (default) `canvas` based rendering of form elements.

Reply:
Thank you very much for the worker import hint, I replaced it and it worked just fine.

I'm 100% ok to split it further, but could you guys explain a little bit problem you referred to? Should I add the worker task I use for parsing annotations to the WorkerTasks collection?

Reply:
> I'd thus suggest splitting this up a lot more, for example let's start with a PR that only makes the annotation parsing async, then one to parse the AcroForm dictionary[1], then one to use that dictionary when parsing annotations, and so on...

I think that sounds good. The smaller the patch, the easier it is to verify. I'm not too familiar with how `WorkerTask`s work, unfortunately, so I too could use some additional feedback here. The only reason we're even dealing with a `WorkerTask` is because it's a necessary parameter. What should we do differently with it?